### ISSUE TITLE
[Bug] Wrapping a DistributedDataParallel module with another One leads to incorrect behavior.

### ISSUE THREAD

**Issue by Gao-HaoYuan** [pytorch/pytorch#170678]
### 🐛 Describe the bug

I encountered unexpected behavior when wrapping a PyTorch nn.Module with two layers of DistributedDataParallel (DDP), i.e. creating a DDP instance on top of another DDP-wrapped module.

While this is not a documented or supported use case, the current behavior is non-obvious and can lead to runtime   incorrect gradient synchronization on a single-node, multi-GPU setup. I’m reporting this to clarify whether this usage should be explicitly forbidden, detected, or better documented.

### Versions

The Code Demo below shows
```   
ddp1 = DDP(
        model,
        device_ids=[local_rank],
        output_device=local_rank,
        find_unused_parameters=True
    )

ddp2 = DDP(
    ddp1,
    device_ids=[local_rank],
    output_device=local_rank,
    find_unused_parameters=True
)
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

**Comment by Gao-HaoYuan** [comment]
@mikaylagawarecki @wwwjn @weifengpy @skpark-rh 

Sorry for the confusion and for taking your time.

After further debugging, I found that the issue was caused by my own mistake in checkpoint loading, not by DDP itself. This was a result of our rushed assumption, and I apologize for the unnecessary noise it caused.

The reason is:

When the model is wrapped with nested (double-wrapped) DistributedDataParallel, the state_dict parameter names end up having an extra "module." prefix compared to the single-wrapped setup. In our code path that reuses checkpoints, it were loading with strict=False. As a result, the keys did not match, most of the checkpoint weights were not actually loaded, and the run proceeded with (partially) randomly initialized weights. This manifested as “inconsistent results” and led me to suspect nested DDP.

After fixing the checkpoint key mapping / loading logic (ensuring the correct prefix handling and verifying the number of matched keys), the results are consistent. I also ran multiple additional experiments with nested DDP, and in all of them the final results can be matched when the checkpoint is correctly reused.

I am very sorry for the confusion and for taking your time — this was an issue on our side.

I am very sorry!

**Comment by weifengpy** [comment]
@Gao-HaoYuan no worreis. you raised a good point. good to know nested ddp actually work
